src/test/tuntap.cpp

```diff
@@ -152,7 +155,128 @@ TEST(tuntap, read_from_tun) {
 		tuntap.read_from_tun(receive_buffer.data(), receive_buffer.size())
 		,0U
 	);
+}
+
+TEST(tuntap, read_from_tun_separated_addresses) {
+	c_tuntap_linux_obj tuntap;
+	auto normal_process =
+	[](const std::array<boost::asio::mutable_buffer, 4> &bufs) -> size_t {
+		std::array<unsigned char, IPV6_LEN> src, dst;
+		std::array<unsigned char, 71> packet;
+		std::tie(src, dst, packet) = get_separated_packet();
+		// check buffers siazes
+		EXPECT_EQ(boost::asio::buffer_size(bufs.at(0)), 8U);
+		EXPECT_EQ(boost::asio::buffer_size(bufs.at(1)), IPV6_LEN);
+		EXPECT_EQ(boost::asio::buffer_size(bufs.at(2)), IPV6_LEN);
+		EXPECT_EQ(boost::asio::buffer_size(bufs.at(3)), 71U - 8);
+		// copy data into buffers
+		boost::asio::buffer_copy(bufs.at(0), boost::asio::buffer(packet.data(), 8));
+		boost::asio::buffer_copy(bufs.at(1), boost::asio::buffer(src.data(), src.size()));
+		boost::asio::buffer_copy(bufs.at(2), boost::asio::buffer(dst.data(), dst.size()));
+		boost::asio::buffer_copy(bufs.at(3), boost::asio::buffer(packet.data()+8, packet.size()-8));
+		return src.size() + dst.size() + packet.size();
+	};
+	auto error_process =
+	[](const std::array<boost::asio::mutable_buffer, 4> &bufs) -> size_t {
+		_UNUSED(bufs);
+		throw boost::system::system_error(boost::asio::error::eof);
+	};
+
+	EXPECT_CALL(tuntap.m_tun_stream, read_some(An<const std::array<boost::asio::mutable_buffer, 4>&>()))
```

Comment:
// We define that the tests will result in correct read on call number 1 and 2,  and will fail on call 3 and 4.
// below with EXPECT_NO_THROW etc we run this mock'ed scenario